### PR TITLE
Turn off automatic conversion of locals to args in CUDA

### LINKED ISSUE #823: CUDA: constant size local memory is allocated as dynamic shared memory
If I compile the following kernel using POCL and inspect the PTX, the OpenCL local memory is declared extern (dynamic shared memory in CUDA). If I compile the kernel with the same clang version (9.0.1) using `clang -x cl -S kernel.cl -o kernel.ptx --target=nvptx64-unknown-unknown  -Xclang -finclude-default-header`, then the local memory is declared as static shared memory in CUDA. The dynamic shared memory kernel is slow as it leads to lesser optimization opportunities.

cc @nchristensen, @inducer, @jrprice 

```OpenCL
#define lid(N) ((int) get_local_id(N))
#define gid(N) ((int) get_group_id(N))

__kernel void __attribute__ ((reqd_work_group_size(32, 32, 1))) diff(__global float *__restrict__ result, __global float const *__restrict__ mat, __global float const *__restrict__ vec)
{
    float acc_j;
    __local float matf[32 * 32];
    __local float vecf[32 * 32];

    vecf[32 * lid(1) + lid(0)] = vec[1048576 * lid(1) + 32 * gid(0) + lid(0)];
    matf[32 * lid(1) + lid(0)] = mat[32 * lid(1) + lid(0)];
    acc_j = 0.0f;
    barrier(CLK_LOCAL_MEM_FENCE)
    /* for matf (insn_j_update depends on mat_fetch_rule) */;
    for (int j = 0; j <= 31; ++j)
        acc_j = acc_j + matf[32 * lid(1) + j] * vecf[32 * j + lid(0)];
    result[1048576 * lid(1) + 32 * gid(0) + lid(0)] = acc_j;
}
```

### PR DESCRIPTION
**Pull request by isuruf**
Turning it on prevents optimization opportunities.

Fixes #823

### CODE REVIEW

**Comment by pjaaskel** [comment]
I suppose there were no regressions after switching this and you tested with different numbers of OpenCL locals, dynamic and static? I do not have a CUDA env set up to test at this moment.

**Comment by isuruf** [comment]
There are some regressions. I'll have a look and mark this PR as ready when done.

**Comment by isuruf** [comment]
Here's what's this patch is doing.
1. Disables converting constant sized __local arrays to the extern block if there are no __local arguments.
2. If there are __local arguments, local arrays with constant size is stored at the beginning of the `extern` block with constant offsets. Dynamic sized ones come later and they are indexed with a constant + unsigned dynamic offset.

This helps with what I think is a pointer aliasing issue.

**Comment by Oblomov** [comment]
I haven't looked at the code yet, but from your last comment, the constant + dynamic offset sounds like the correct solution.

In CUDA, all `extern __shared__` memory blocks have the same base pointer address, because they point to the area allocated by the third parameter in the CUDA kernel launch syntax `kernelName<<<numBlocks, threadsPerBlock, dynShMemSizeInBytes, commandStream>>>(arguments). So yeah, it's not possible to convert `__local` arguments (i.e. dynamically sized LDS) in OpenCL to `extern __shared__` “as is”: instead, some kind of transformation has to happen, and the offset cannot be constant because it depends on the sizing of the other dynamically allocated memory blocks (taking pointer alignment into consideration, of course).

**Comment by pjaaskel** [comment]
@jrprice does it look OK to you? I'm trying to setup pocl-cuda in my box so I can test it. Is there an expected working test suite I can run?

**Comment by pjaaskel** [comment]
Can you @isuruf update CHANGES please? Preferably with some speedup estimates from this update, which we can advertise in the next release.

**Comment by pjaaskel** [comment]
Also please try format-last-commit.sh, I see some suspicious style issues.

File: CHANGES [hunk 1]
```diff
@@ -1,6 +1,47 @@
...
 ==============
 
+Notable User Facing Changes
+---------------------------
+
+- CUDA kernels using constant __local blocks are now ABI incompatible with
+  previous release. Users need to delete their pocl cache.
+
+
+Notable Bug Fixes
```
Comment (pjaaskel) [inline CHANGES:11]: Is that really a bug fix? I'd put it under Optimizations or such title. Also can you summarize the performance improvement in terms of percentages instead of listing all the benchmarks. Great work!
